### ISSUE TITLE
Alternative ideas

### ISSUE THREAD

**Issue by davidtheclark** [mapbox/scroll-restorer#1]
@jfurrow: In https://github.com/mapbox/batfish/commit/e52a0226ee70bcf2744f029d33b25951dee5120e I experimented with some changes to this script, spurred by an annoying experience in IE10. Here were a few of my ideas:

- I think the "attempts" system for `syncScroll` wasn't working properly because of the way `debounce` was applied (like, the second attempt never fired because it got debounced?). I came to think that `syncScroll` doesn't *need* to be debounced, because it's not tied to a scroll event.
- I wonder if `requestAnimationFrame` is enough debounce on the scroll event that we can remove the actual `debounce` from `captureScroll`. The original debounce of 50ms is almost certainly too small to make much difference over `requestAnimationFrame`, right?
- The way Batfish's router is working, I thought that maybe `syncScroll` might try to run *before* the page component had actually changed — not ideal. So I experimented with exposing that function as part of the API (changed name to `restoreScroll`) and having Batfish's router call it strategically, after `setState` had finished changing the page component.
- A warning: IE10 is messy with this stuff. It tries to do something on its own that gets in the way. But I think the above might tame it. And it's not *that* important, especially if it *works* but is just a little bouncy sometimes.

The way Batfish is currently configured seems to work in IE10 & 11 and our good friends Chrome, Firefox, and Safari.

Once you have a chance to dig in, see what you think of those changes! We'll eventually change Batfish to fit whatever you think works best — these are just some more ideas I came up with today.

**Comment by jfurrow** [comment]
Thanks for the overview, @davidtheclark! I'll be sure to consider these findings as I dive in.

**Comment by jfurrow** [comment]
I agree that `syncScroll`/`restoreScroll` doesn't need to be debounced, because as you said, it's not tied to the `scroll` event. And, as you pointed out, when it _is_ called repeatedly, we want it to be invoked immediately.

I can't think of a good reason _not_ to `debounce` `captureScroll`. We only need to store the scroll position when the user has finished scrolling, so I think using both `debounce` and `requestAnimationFrame` would be best.

I'm still testing these changes locally, so I'll follow up with more feedback when I've fully implemented your changes in my example app.

**Comment by jfurrow** [comment]
@davidtheclark I implemented your changes here: https://github.com/mapbox/scroll-restorer/pull/3. I went ahead and modified it to `debounce` `captureScroll` — are you cool with this? As I said above, I think it provides some value.

I've manually tested this pretty thoroughly and it works great! I think it's time to start writing some automated tests based on the API introduced at https://github.com/mapbox/scroll-restorer/pull/3. Then on to documentation!
